compositor_pipeline/src/pipeline/decoder/audio/opus.rs

```diff
@@ -49,25 +50,99 @@ impl OpusDecoder {
         )
         .into()
     }
+
+    fn set_end_pts(&mut self, decoded_samples: &DecodedSamples) {
+        let samples_len = decoded_samples.samples.get_number_of_samples();
+        let sample_rate = decoded_samples.sample_rate;
+
+        let chunk_duration = Duration::from_secs_f64(samples_len as f64 / sample_rate as f64);
+        self.last_decoded_pts = Some(decoded_samples.start_pts + chunk_duration);
+    }
+
+    fn should_use_fec(&self, stream_gap: Duration) -> bool {
+        stream_gap > Duration::from_millis(1)
+    }
+
+    fn calculate_stream_gap(&mut self, current_start: Duration) -> Duration {
+        let stream_gap = current_start - *self.last_decoded_pts.get_or_insert(current_start);
+        trace!(
+            "[opus decoder] Calculated stream gap: {} s",
+            stream_gap.as_secs_f64(),
+        );
+        stream_gap
+    }
+
+    fn decode_chunk(
+        &mut self,
+        encoded_chunk: &EncodedChunk,
+    ) -> Result<DecodedSamples, DecodingError> {
+        let decoded_samples_count =
+            self.decoder
+                .decode(&encoded_chunk.data, &mut self.decoded_samples_buffer, false)?;
+
+        let samples = Self::read_buffer(&self.decoded_samples_buffer, decoded_samples_count);
+        Ok(DecodedSamples {
+            samples,
+            start_pts: encoded_chunk.pts,
+            sample_rate: self.decoded_sample_rate,
+        })
+    }
+
+    fn decode_chunk_fec(
+        &mut self,
+        encoded_chunk: &EncodedChunk,
+        stream_gap: Duration,
+    ) -> Result<DecodedSamples, DecodingError> {
+        debug!("[opus decoder] FEC used!");
+
+        let lost_samples = self.decoder.get_last_packet_duration()?;
+
+        // 120 samples is 2.5 ms with 48kHz sample rate. For FEC it is mandatory that buffer size
+        // is a multiple of 2.5 ms and of the same size (or at least as close as possible) to the
+        // size of lost chunk
+        let fec_buf_size = 120 * (lost_samples as usize / 120);
```

Comment:
Given that the sample calculation might depend on duration, which might not be numerically precise, I think rounding to the closest multiple of 2.5 might give a better result than the floor